Bitrot and scrub process cmdline arg addition

schaffung opened the pull request:
The bitd and scrubd services when viewed from the
perspective of the `gluster volume status <volname> client-list`
returned `unknown` as the `--process-name` cmdline
arg wasn't provided when starting up these services.

Fixes: #2398
Signed-off-by: srijan-sivakumar <ssivakum@redhat.com>

Commit:
Bitrot and scrub process cmdline arg addition

The bitd and scrubd services when viewed from the
perspective of the `gluster volume status <volname> client-list`
returned `unknown` as the `--process-name` cmdline
arg wasn't provided when starting up these services.

Fixes: #2398
Signed-off-by: srijan-sivakumar <ssivakum@redhat.com>

schaffung commented:
/run regression

schaffung commented:
/recheck smoke

pranithk commented on xlators/mgmt/glusterd/src/glusterd-scrub-svc.c:
```diff
@@ -113,15 +113,25 @@ glusterd_scrubsvc_manager(glusterd_svc_t *svc, void *data, int flags)
+        if (ret)
+            goto dealloc_dict;
+    }
+
```
Both these functions are exactly same. Do you see any reason for them to be separate functions?

schaffung replied:
Now that I think..yes! Both the functions are duplicates. Will make the necessary changes.

schaffung replied:
Also, if we are removing duplicates, glusterd_quotadsvc_start is also similar, the only difference being a certain cmdline args. I think even that should be handled, as we are anyway removing duplicates.

pranithk replied:
glusterd seems to have lot of duplicate code. If you want to send another patch removing more duplicate code, I will be happy to review.

pranithk commented:
Could you add a .t as well?

schaffung commented:
> Could you add a .t as well?

Ack.

Commit:
Modifying an existing bitrot TC

After fixing the output for the client-list output
for bitd and scrub, modified an existing TC to
verify the scenario.

Signed-off-by: srijan-sivakumar <ssivakum@redhat.com>

schaffung commented:
@pranithk have modified an existing .t which already handles bitrot functioning to check the client-list status when bitrot is enabled.

pranithk commented:
> @pranithk have modified an existing .t which already handles bitrot functioning to check the client-list status when bitrot is enabled.

Are you planning to address the code duplication comment in this PR?

Commit:
Removed duplicate functions.

Replaced the functions glusterd_bitdsvc_start and
glusterd_scrubsvc_start with glusterd_genericsvc_start.

Updates: #2398
Signed-off-by: srijan-sivakumar <ssivakum@redhat.com>

schaffung commented:
/run regression

pranithk commented on xlators/mgmt/glusterd/src/glusterd-svc-mgmt.c:
```diff
@@ -517,3 +517,43 @@ glusterd_muxsvc_conn_init(glusterd_conn_t *conn, glusterd_svc_proc_t *mux_proc,
     return ret;
 }
+
+/*
+ * A generic function replacing two functions,
+ * glusterd_bitdsvc_start and glusterd_scrubsvc_start
+ * wherein both do the same set of operations.
+ */
```
The function name you gave to `glusterd_genericsvc_start` already conveys this info.
LGTM anyways.

schaffung replied:
Ack. Just added a note so that we know what this function replaces. I didn't replace the function in quota. Probably will take it as a separate patch and see if there are any more duplicates.

nik-redhat commented on xlators/mgmt/glusterd/src/glusterd-svc-mgmt.c:
```diff
@@ -517,3 +517,43 @@ glusterd_muxsvc_conn_init(glusterd_conn_t *conn, glusterd_svc_proc_t *mux_proc,
+    }
+
+    for (i = 0; options[i]; i++) {
+        ret = snprintf(key, sizeof(key), "arg%d", i);
```
nit: a fail check after this line could be helpful?

schaffung replied:
If the check is for overflow, I think we can ignore it as the size of data being stored in key would be safely within the limits according to the flow of the code here.

nik-redhat commented on xlators/mgmt/glusterd/src/glusterd-svc-mgmt.c:
```diff
@@ -517,3 +517,43 @@ glusterd_muxsvc_conn_init(glusterd_conn_t *conn, glusterd_svc_proc_t *mux_proc,
+
+    ret = glusterd_svc_start(svc, flags, cmdline);
+
+dealloc_dict:
+    dict_unref(cmdline);
+error_return:
+    return ret;
```
Instead of using multiple `goto` sections, we can simply use one like `out`, since we are only going to `error_return` directly only at line: 539, we can return -1 from there itself.

Commit:
Update glusterd-svc-mgmt.c

Simplified the return structure by removing a goto.

schaffung commented:
/run regression
